### ISSUE TITLE
rstan::gqs() not correctly extracting posterior draws

### ISSUE THREAD

**Issue by fweber144** [stan-dev/rstan#695]
#### Summary:
In some cases, `rstan::gqs()` leads to the error `Error in draws[, p_names, drop = FALSE] : subscript out of bounds`.

#### Description:
It seems that when parameters stem from an array or matrix (in Stan code), `rstan::gqs()` is not able to extract them correctly from the matrix that is supplied through argument `draws` of `rstan::gqs()`. To me, it seems like the lines
`p_names <- unique(sub("\\..*$", "", sampler$constrained_param_names(TRUE, 
                                                                      FALSE)))`
`draws <- draws[, p_names, drop = FALSE]`
`all_names <- sampler$constrained_param_names(TRUE, TRUE)`
`some_names <- sampler$constrained_param_names(TRUE, FALSE)`

in the source code of `rstan::gqs()` would have to be replaced by something like:

`p_names <- unique(sub("\\..*$", "", sampler$constrained_param_names(TRUE, 
                                                                      FALSE)))`
`all_names <- sampler$constrained_param_names(TRUE, TRUE)`
`some_names <- sampler$constrained_param_names(TRUE, FALSE)`
`some_names_mod <- sub("\\.", "[", some_names)`
`some_names_mod <- sub("\\.", ",", some_names_mod)`
`some_names_mod[grep("\\[", some_names_mod)] <- paste0(some_names_mod[grep("\\[", some_names_mod)], "]")`
`draws <- draws[, some_names_mod, drop = FALSE]`

#### Reproducible Steps:
Modified code from the "example" section of `rstan::gqs()`:

`library(rstan)`
`m <- stan_model(model_code = 'parameters {real y[3];} model {y ~ normal(0,1);}')`
`f <- sampling(m, iter = 300)`
`mc <- 'parameters {real y[3];}`
`generated quantities {real y_rep[3] = normal_rng(y, 1);}'`
`m2 <- stan_model(model_code = mc)`
`f2 <- gqs(m2, draws = as.matrix(f))`

This returns:
`Error in draws[, p_names, drop = FALSE] : subscript out of bounds`

#### Current Output:
An error message (see above).

#### Expected Output:
Generated quantities for `y[1]`, `y[2]`, `y[3]`, i.e. in this case an S-by-3 matrix (whereby "S" denotes the number of posterior draws).

#### RStan Version:
2.19.2

#### R Version:
R version 3.6.1 (2019-07-05)

#### Operating System:
Windows 10

**Comment by jgabry** [comment]
Thanks for the detailed report. One of us will investigate and get this sorted out for the next release.

**Comment by jgabry** [comment]
@bgoodri I think this is definitely a bug. gqs() only seems to work for scalars at the moment. 

Here's a runnable example that fails with a basic `vector` parameter. gqs() also fails with the same error for matrices and other types (e.g. arrays as reported above) other than scalars.  

```r
m <- stan_model(model_code = 'parameters {vector[2] y;} model {y ~ normal(0,1);}')
f <- sampling(m, iter = 300)

mc <-
'
parameters {
  vector[2] y;
}
generated quantities {
  vector[2] y_rep; 
  for (i in 1:2) y_rep[i] = normal_rng(y[i], 1);
}
'
m2 <- stan_model(model_code = mc)
f2 <- gqs(m2, draws = as.matrix(f))

Error in draws[, p_names, drop = FALSE] : subscript out of bounds
```

**Comment by jgabry** [comment]
@mki71 Thanks for suggesting the fix. It seems correct except

> some_names_mod <- sub("\\.", ",", some_names_mod)

needs to be `gsub` so it replaces all cases (otherwise arrays with 3 or more dimensions still will have "."). I've made that tweak and I think this is now fixed by PR #698.
